packages/core/src/browser/widgets/widget.ts

```diff
@@ -46,11 +46,15 @@ export class BaseWidget extends Widget {
     readonly onScrollUp: Event<void> = this.onScrollUpEmitter.event;
     protected readonly onDidChangeVisibilityEmitter = new Emitter<boolean>();
     readonly onDidChangeVisibility = this.onDidChangeVisibilityEmitter.event;
+    protected readonly onDidDisposeEmitter = new Emitter<void>();
+    readonly onDidDispose = this.onDidDisposeEmitter.event;
 
     protected readonly toDispose = new DisposableCollection(
         this.onScrollYReachEndEmitter,
         this.onScrollUpEmitter,
-        this.onDidChangeVisibilityEmitter
+        this.onDidChangeVisibilityEmitter,
+        this.onDidDisposeEmitter,
```

Comment:
these lines should be first to add to be run the last after all emitters and listeners are released